src/startup.S

```diff
@@ -0,0 +1,44 @@
+.cfi_sections .debug_frame
+
+# .thumb
+.section .text.pre_init, "ax"
+.globl __pre_init
+.type __pre_init,%function
+.thumb_func
+.cfi_startproc
+__pre_init:
+
+  # Enable ITCM and DTCM
+  ldr r0, =1
+  ldr r1, =0xE000EF90
+  ldr r2, [r1]
+  # Set ITCMCR.EN
+  orr r2, r2, r0
+  str r2, [r1]
+  ldr r1, =0xE000EF94
+  ldr r2, [r1]
+  # Set DTCMCR.EN
+  orr r2, r2, r0
+  str r2, [r1]
+  dsb
+  isb
+
+  # Analogous to cortex-m-rt Reset code for .data copying.
+  # Initialise .itcm code. `__sitcm`, `__siitcm`, and `__eitcm` come from the
+  # linker script. Copy from r2 into r0 until r0 reaches r1.
+  ldr r0,=__sitcm
```

Comment:
General comment, but if we can guarantee that no ITCM code executes before `hardware::setup()`, why don't we do this initialization in rust-land? Then we can avoid the whole startup code / assembly section entirely. It may simplify things?

Reply:
Better yet, we could use a [`cortex-m-rt::pre_init` hook](https://docs.rs/cortex-m-rt/0.6.13/cortex_m_rt/attr.pre_init.html)

Reply:
I think that simplifies the build, since we don't need cc or any of the toolchain available

Reply:
This is the `pre_init` hook.
Doing it in rust isn't completely trivial. I'd rather get a working solution first and then later optimize it.
https://github.com/rust-embedded/cortex-m-rt/issues/300 and https://github.com/rust-embedded/cortex-m-rt/issues/301

Reply:
Very valid point - that pointer provenance issue is super annoying.

Reply:
Did it anyway. Seems to work.

Reply:
The problem is that it's _technically_ undefined behavior, so it _might_ break in the future - rust doesn't guarantee anything. It worked for years on cortex-m-rt, but then they noticed it and removed it. I don't know if it actually had a problem in that case, but a user was trying to reproduce what they had done and couldn't get it to work.

Reply:
I don't think it is in this case. c-m-rt used `ptr` because it has no stack at that point. The problem is in using a `ptr::offset`.
Here I'm creating and using slices the same way you'd do for FFI. That also works "out of thin air".

Reply:
I think we want to do what they did in https://github.com/rust-embedded/r0/pull/25 to get around this.

Reply:
There are already two.

Reply:
And `r0` isn't used anymore.